test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,202 @@
+{
+	_values.clear();
+	for (string const& testId: m_expectationsSequence)
+	{
+		soltestAssert(m_testCases.count(testId) > 0);
+		_values +=
+			testId +
+			": " +
+			(_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+			+ "\n";
+	}
+}
+
+vector<StringPair> ASTPropertyTest::readKeyValuePairs(string const& _input)
+{
+	vector<StringPair> result;
+	for (string const& line: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		if (line.empty())
+			continue;
+
+		vector<string> pair;
+		for (string const& element: line | ranges::views::split(':') | ranges::to<vector<string>>)
+		{
+			soltestAssert(ranges::all_of(boost::trim_copy(element), [](char c) { return isprint(c); }));
+			pair.emplace_back(boost::trim_copy(element));
+		}
+
+		soltestAssert(pair.size() == 2, "Invalid property test: " + line);
+		result.emplace_back(pair[0], pair[1]);
+	}
+	return result;
+}
+
+void ASTPropertyTest::readExpectations()
+{
+	for (auto const& [testId, testExpectation]: readKeyValuePairs(m_reader.simpleExpectations()))
+	{
+		soltestAssert(m_testCases.count(testId) == 0, "More than one expectation for test \"" + testId + "\"");
+		m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+		m_expectationsSequence.push_back(testId);
+	}
+	generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+	if (_root.isObject() && !_property.empty())
+	{
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+	queue<Json::Value> nodesToVisit;
+	nodesToVisit.push(_astJson);
+
+	set<string> readTests;
+
+	while (!nodesToVisit.empty())
+	{
+		Json::Value& node = nodesToVisit.front();
+
+		if (node.isArray())
+			for (auto&& member: node)
+				nodesToVisit.push(member);
+		else if (node.isObject())
+			for (string const& memberName: node.getMemberNames())
+			{
+				if (memberName != "documentation")
+				{
+					nodesToVisit.push(node[memberName]);
+					continue;
+				}
+
+				string nodeDocstring = node["documentation"].isObject() ?
+					node["documentation"]["text"].asString() :
+					node["documentation"].asString();
+				soltestAssert(!nodeDocstring.empty());
+
+				vector<StringPair> pairs = readKeyValuePairs(nodeDocstring);
+				soltestAssert(!pairs.empty(), "It is required at least one property test per annotated AST node.");
+
+				for (auto const& [testId, testedProperty]: pairs)
+				{
+					soltestAssert(
+						readTests.insert(testId).second,
+						"Test \"" + testId + "\" was already defined before."
+					);
```

Comment:
I'd avoid asserts with side-effects as a general rule.

It's less of a problem with our asserts than with the standard `assert()` macro, since our asserts are always executed, but still, it feels just wrong and while reading code I rarely expect such code to actually do anything other than a simple side-effect-free check.
```suggestion
					bool inserted = readTests.insert(testId).second;
					soltestAssert(inserted,	"Test \"" + testId + "\" is already defined.");
```

Reply:
By the way, `readTests` actually seems a bit superfluous. You could disallow tests with empty property path (they are not useful anyway) and determine if the test has been defined by checking if `property` field is empty.

Reply:
Yeah, I removed `readTests` and now the code checks `property` field as you suggested.